### PR TITLE
Sort Section Search Results by Subscription then Open Seat Count

### PR DESCRIPTION
**Pull request by soulwa**
# Purpose

Currently, sections are sorted by professor names. It might be more helpful for users to sort by subscription and open seat count.

<br>

# Tickets

###### A link to any tickets this PR is associated with on Trello.

- https://trello.com/c/6ULqeoqP

# Contributors

###### Anyone who contributed to this PR for future reference.

- @soulwa 
- @hankewyczz 

# Feature List

###### A list of more in-depth and technical changes, additions, deletions, and fixes this PR provides.

- replace `sortSections` with a sort by subscription status, then remaining section seats, then remaining waitlist seats in `CourseResult.tsx`, have it take `userInfo` as necessary context for sorting by subscription.

# Checklist

- [x] Filled out PR template :wink:
- [ ] Approved by designers
- [ ] Is passing linting checks
- [ ] Is passing tsc
- [ ] Is passing existing tests
- [ ] Has documentation and comments in code
- [ ] Has test coverage for code
- [ ] Will have clear squash commit message

<br>
@sandboxnu/searchneu

### CODE REVIEW

File: components/ResultsPage/Results/CourseResult.tsx [hunk 1]
```diff
@@ -48,7 +57,10 @@ export function CourseResult({
   const router = useRouter();
   const termId = router.query.termId as string;
   const campus = router.query.campus as string;
-  const sortedSections = useMemo(() => sortSections(course.sections), [course]);
+  const sortedSections = useMemo(
+    () => sortSections(course.sections, userInfo),
+    [course]
```
Comment (hankewyczz) [inline components/ResultsPage/Results/CourseResult.tsx:61]: We should add `userInfo` as a dependency (and the same for the lower `useMemo` instance)

Comment (soulwa) [reply components/ResultsPage/Results/CourseResult.tsx:61]: iirc we didn't do this because this would make classes which you registered for immediately appear at the top of the list (it would resort the list right away, rather than on refresh) -- is this behavior we want for the UI?

Comment (hankewyczz) [reply components/ResultsPage/Results/CourseResult.tsx:61]: ahh, sorry, I forgot about that. You're right

File: components/ResultsPage/Results/CourseResult.tsx [hunk 2]
```diff
@@ -17,24 +17,33 @@ import { DesktopSectionPanel, MobileSectionPanel } from './SectionPanel';
...
   onSignIn: (token: string) => void;
   fetchUserInfo: () => void;
 }
 
-const sortSections = (sections: Section[]): Section[] => {
+const sortSections = (sections: Section[], userInfo?: UserInfo): Section[] => {
+  // NOTE (sam 2023-01-29): unsure why sections needs to be cloned, because all we're doing is sorting, nothing desctructive.
+  // also unsure why sections are memoized after being sorted (in CourseResult)
+  // memoized based on current course, but course should never change over lifetime of a CourseResult
   const sortedSections = cloneDeep(sections);
```
Comment (Lucas-Dunker) [inline components/ResultsPage/Results/CourseResult.tsx]: If we're unsure, would removing the clone be a separate PR or could we just include it in this one? Same for the memorization

Comment (soulwa) [reply components/ResultsPage/Results/CourseResult.tsx]: alright so- 
`cloneDeep` is necessary here because of `useMemo`, since `useMemo` require that its function argument be pure. *technically* we could get away with a shallow copy here, since this sort isn't (and never should be?) modifying the actual `Section` objects in the array. but for safety I'm not opposed to leaving the `cloneDeep`, as long as we leave the `useMemo`. I'm open to leaving it in at the moment in case we want to have course results reorder themselves on notification, since that is not a behavior we want now, but could maybe do something with in the future. 

so options are
1. leave `useMemo` and `cloneDeep` (guarantees pure)
2. leave `useMemo`, move to shallow copy (have to uphold invariant that sort doesn't modify objects, which it **shouldn't**, but also removes a `lodash` dependency)

**Review by Lucas-Dunker — APPROVED** [review]

**Comment by soulwa** [comment]
finally had a chance to actually look into this `useMemo` business- I don't think it's necessary, but right now without it `useShowAll` causes hundreds of re-renders. I've got an idea for a way to decouple this, but to get this PR through I'm going to update the comment and leave it be.

**Review by pranavphadke1 — APPROVED** [review]
